include/kitty/decomposition.hpp

```diff
@@ -237,6 +249,533 @@ bottom_decomposition is_bottom_decomposable( const TT& tt, uint32_t var_index1,
+      {
+        xb.push_back( h );
+      }
+    }
+    else
+    {
+      auto xa_p = xa;
+      auto xb_p = xb;
+      xa_p.push_back( h );
+      xb_p.push_back( h );
+      if ( check_or_decomp( tt, dc, xa, xb_p ) )
+      {
+        xb.push_back( h );
+      }
+      else if ( check_or_decomp( tt, dc, xa_p, xb ) )
+      {
+        xa.push_back( h );
+      }
+    }
+  }
+  return ( std::make_pair( xa, xb ) );
+}
+
+template<class TT>
+std::pair<std::vector<int>, std::vector<int>> group_variables_weak_or( const TT& tt, const TT& dc, const std::vector<int>& s )
+{
+  std::vector<int> xa, xb;
+  auto var_x = find_initial_weak_or( tt, dc, s );
+  xa = var_x.first;
+  xb = var_x.second;
+  if ( ( xa.size() == 0 ) && ( xb.size() == 0 ) )
+  {
+    return var_x;
+  }
+  for ( auto h : s )
+  {
+    auto it = find( xa.begin(), xa.end(), h );
+    if ( it != xa.end() )
+    {
+      continue;
+    }
+    auto xa_p = xa;
+    xa_p.push_back( h );
+    if ( check_weak_decomp( tt, dc, xa_p ) )
+    {
+      xa.push_back( h );
+    }
+  }
+  return ( std::make_pair( xa, xb ) );
+}
+
+template<class TT>
+std::tuple<std::vector<int>, std::vector<int>, std::vector<TT>> group_variables_xor( const TT& tt, const TT& dc, const std::vector<int>& s )
+{
+  std::vector<int> xa, xb;
+  std::vector<TT> q_and_r;
+  auto var_x = find_initial_xor( tt, dc, s );
+  xa = std::get<0>( var_x );
+  xb = std::get<1>( var_x );
+  q_and_r = std::get<2>( var_x );
+  if ( ( xa.size() == 0 ) && ( xb.size() == 0 ) )
+  {
+    return var_x;
+  }
+  for ( auto h : s )
+  {
+    auto it = find( xa.begin(), xa.end(), h );
+    if ( it != xa.end() )
+    {
+      continue;
+    }
+    it = find( xb.begin(), xb.end(), h );
+    if ( it != xb.end() )
+    {
+      continue;
+    }
+
+    if ( xa.size() <= xb.size() )
+    {
+      auto xa_p = xa;
+      auto xb_p = xb;
+      xa_p.push_back( h );
+      xb_p.push_back( h );
+      auto f = check_xor_decomp( tt, dc, xa_p, xb );
+      auto f2 = check_xor_decomp( tt, dc, xa, xb_p );
+      if ( f.first )
+      {
+        xa.push_back( h );
+        q_and_r = f.second;
+      }
+      else if ( f2.first )
+      {
+        xb.push_back( h );
+        q_and_r = f2.second;
+      }
+    }
+    else
+    {
+      auto xa_p = xa;
+      auto xb_p = xb;
+      xa_p.push_back( h );
+      xb_p.push_back( h );
+      auto f = check_xor_decomp( tt, dc, xa_p, xb );
+      auto f2 = check_xor_decomp( tt, dc, xa, xb_p );
+      if ( f2.first )
+      {
+        xb.push_back( h );
+        q_and_r = f2.second;
+      }
+      else if ( f.first )
+      {
+        xa.push_back( h );
+        q_and_r = f.second;
+      }
+    }
+  }
+  return ( std::make_tuple( xa, xb, q_and_r ) );
+}
+
+std::tuple<std::vector<int>, std::vector<int>, bi_decomposition> best_variable_grouping( std::pair<std::vector<int>, std::vector<int>> x_or, std::pair<std::vector<int>, std::vector<int>> x_and, std::pair<std::vector<int>, std::vector<int>> x_xor )
```

Comment:
All arguments should be passed const reference, unless they are changed in the code.